File: Leetcode_Code/95. Unique Binary Search Trees II/95.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<TreeNode *> generateTrees(int n) {
        vector<TreeNode *> err;
        if (n < 1) return err;
        return generateSubTree(1, n);
    }

private:
    vector<TreeNode *> generateSubTree(int start, int end) {
        vector<TreeNode *> result;
        if (end < start) {
            result.push_back(NULL);
            return result;
        }
        for (int i = start; i <= end; ++i) {
            for (TreeNode *subL: generateSubTree(start, i - 1)) {
                for (TreeNode *subR:generateSubTree(i + 1, end)) {
                    TreeNode *head = new TreeNode(i);
                    head->left = subL;
                    head->right = subR;
                    result.push_back(head);
                }
            }
        }
        return result;
    }
};

File: Leetcode_Code/95. Unique Binary Search Trees II/95.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<TreeNode> generateTrees(int n) {
        if (n < 1) {
            return new ArrayList<>();
        }
        return generateSubTrees(1, n);
    }

    private List<TreeNode> generateSubTrees(int start, int end) {
        List<TreeNode> result = new ArrayList<>();
        if (end < start) {
            result.add(null);
            return result;
        }
        for (int mid = start; mid <= end; mid++) {
            for (TreeNode SubL : generateSubTrees(start, mid - 1)) {
                for (TreeNode SubR : generateSubTrees(mid + 1, end)) {
                    TreeNode root = new TreeNode(mid);
                    root.left = SubL;
                    root.right = SubR;
                    result.add(root);
                }
            }
        }
        return result;
    }
}
